Cce analytic test teukolsky wave

moxcodes opened the pull request:
## Proposed changes

Add the Teukolsky wave analytic solution for CCE

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] PR #2261
- [x] PR #2287

moxcodes committed:
Add TeukolskyWave analytic solution

kidder commented:
could you factor out the first two commits, we can get those in fast...

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,373 @@
+  static DataVector cos_theta(size_t l_max) noexcept;
+
+ protected:
+  /// A no-op as the gauge wave solution does not have substantial
```
should gauge wave -> Teukolsky wave?

fmahebert replied:
ping on this one (should gauge -> Teukolsky?)

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,373 @@
+   * F(u) = a e^{-u^2 /k^2}.
+   * \f]
+   *
+   * \note The \f$\phi\f$ components are returned in 'pfaffian' form, where the
```
[nitpick] pfaffian -> Pfaffian (also below)

kidder replied:
I would suggest omitting Pfaffian in the documentation because many people don't know what it is.  Just say that we are coding up g_phi_phi with the appropriate power of sin(theta) as these factors typically cancel when multiplied by Jacobians that also are expressed with similarly written terms

fmahebert commented on tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_TeukolskyWave.cpp:
```diff
@@ -0,0 +1,297 @@
+  const double extraction_radius = radius_dist(gen);
+  const size_t l_max = 16;
+  // use a low frequency so that the dimensionless parameter in the metric is ~1
+  const double amplitude = parameter_dist(gen) / 10.0;
```
If I'm reading the code right, this line is the only place you use `parameter_dist`. So why not draw from `{0.05, 0.2}` and avoid dividing by 10?

fmahebert reviewed (commented):
I did not do a detailed review of the math, but this looks okay to me.

Here's three nitpicks; if you make any changes please squash them in directly.

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/BouncingBlackHole.hpp:
```diff
@@ -69,9 +69,8 @@ struct BouncingBlackHole : public WorldtubeData {
-
   // clang doesn't manage to use = default correctly in this case
-  // NOLINTNEXTLINE(hicpp-use-equals-default)
+  // NOLINTNEXTLINE(hicpp-use-equals-default,modernize-use-equals-default)
```
Did you check if the first warning still applies? I've noticed some cases where the newer (llvm 10) clang-tidy has started flagged the same line of code under a different warning

moxcodes commented:
fixups posted.
Cheers

fmahebert commented:
Fixups LGTM, you can squash

moxcodes commented:
squashed.
Cheers!

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,373 @@
+      size_t l_max, double time) const noexcept override;
+
+  /*!
+   * \brief Compute the radial derivative of the spherical coordinate metric
```
radial -> time

nilsdeppe replied:
ping

moxcodes replied:
oops, sorry -- got it now.

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,373 @@
+   * in this analytic solution.
+   */
+  void variables_impl(
+      gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, -2>>*> News,
```
News -> news

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,326 @@
+    noexcept {
+  const double retarded_time = time - extraction_radius_;
+  return 2.0 *
+         (-3.0 * pow<6>(duration_) +
```
this does not agree with the documentation, the leading term should be 1/R^5 according to the doc

kidder replied:
can you make the code and documentation factored the same so it is easier to visually compare?

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,326 @@
+  get<3, 3>(*spherical_metric) =
+      (1.0 + (3.0 * square(local_sin_theta) - 1.0) * coefficient_a -
+       3.0 * square(local_sin_theta) * coefficient_c) *
+      square(extraction_radius_);
```
missing factor sin^2

moxcodes replied:
This one is intentional -- the sin^2 theta is omitted because of the 'pfaffian' form

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,326 @@
+  get<2, 3>(*dr_spherical_metric) = 0.0;
+
+  get<1, 2>(*dr_spherical_metric) =
+      -3.0 * local_sin_theta * local_cos_theta *
```
off by factor of 2 from documentation

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,326 @@
+  get<1, 1>(*dt_spherical_metric) = 0.0;
+  get<1, 3>(*dt_spherical_metric) = 0.0;
+  get<2, 3>(*dt_spherical_metric) = 0.0;
+  get<1, 2>(*dt_spherical_metric) = -3.0 * local_sin_theta * local_cos_theta *
```
off by factor of 2 from doc

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,326 @@
+      std::complex<double>{6.0, 0.0} * square(local_sin_theta) *
+      (15.0 * pow<4>(duration_) -
+       20.0 * square(duration_ * (time - extraction_radius_)) +
+       4.0 * (time - extraction_radius_)) *
```
should be pow<4>(t-r)

kidder commented on tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_TeukolskyWave.cpp:
```diff
@@ -0,0 +1,297 @@
+  const size_t l_max = 16;
+  // use a low frequency so that the dimensionless parameter in the metric is ~1
+  const double amplitude = parameter_dist(gen) / 10.0;
+  const double duration = 1.0;
```
don't use 1.0.  1^5=1^-2477

kidder commented on tests/Unit/Evolution/Systems/Cce/AnalyticSolutions/Test_TeukolskyWave.cpp:
```diff
@@ -0,0 +1,297 @@
+      6.0 * amplitude *
+      (3.0 + square(extraction_radius) *
+                 (-6.0 + extraction_radius * (3.0 + 2.0 * extraction_radius))) /
+      (M_E * pow<6>(extraction_radius)) * sin_theta * cos_theta;
```
also setting k=u is bad for testing;  this seems ideal for testing with python...

kidder reviewed (changes requested):
reviewed with @nilsdeppe

moxcodes commented:
Fixups posted.
Thanks for the reviews!

moxcodes commented:
Fixups posted.
Thanks for the reviews!

kidder commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -76,45 +75,42 @@ double TeukolskyWave::pulse_profile_coefficient_c(const double time) const
   const double retarded_time = time - extraction_radius_;
   return -dt_pulse_profile_coefficient_a(time) -
-         9.0 *
+         9.0 * amplitude_ * exp(-square(retarded_time) / square(duration_)) /
```
can you please write out the derivatives in the documentation so we can compare these easily

kidder reviewed (commented):
you can squash

moxcodes commented:
fixups squashed in along with the extra dox for the derivatives of the pulse profile coefficients.
Thanks! I appreciate the detailed math checks -- those will save me or someone else a lot of trouble down the line.

nilsdeppe commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.cpp:
```diff
@@ -0,0 +1,316 @@
+              8.0 * pow<3>(extraction_radius_) * pow<3>(retarded_time) +
+              6.0 * square(duration_) * square(extraction_radius_) *
+                  retarded_time *
+                  (2.0 * extraction_radius_ + 3.0 * retarded_time) -
```
the documentation has a `+` sign for the last term

moxcodes replied:
Checked through and it was the documentation in error this time -- I squashed the sign fix in. Thanks!

nilsdeppe reviewed (commented):
LGTM, just one inconsistency that I noticed. Feel free to squash that in :)

kidder commented:
not sure what happened with the commits here...

moxcodes commented:
Yeah... weird, sorry -- I can't work out from the reflog what went wrong.
I removed the unintended changes from the 'clang-tidy' commit, and found parts of the analytic data helpers that are no longer used for the python version of the test and dropped those as well.

kidder commented:
@fmahebert please look at

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/BouncingBlackHole.hpp:
```diff
@@ -69,7 +69,8 @@ struct BouncingBlackHole : public WorldtubeData {
   // clang doesn't manage to use = default correctly in this case
-  BouncingBlackHole() noexcept {}  // NOLINT
+  // NOLINTNEXTLINE(modernize-use-equals-default)
+  BouncingBlackHole() noexcept {};
```
This is a surprising semi-colon. Am I missing something, or can it be removed?

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,418 @@
+/// \endcond
+
+namespace Cce {
+namespace Solutions {
```
[not a change request] I'm surprised that clang-tidy isn't complaining about the non-concatenated namespaces here, the way it has been elsewhere...

fmahebert commented on src/Evolution/Systems/Cce/AnalyticSolutions/TeukolskyWave.hpp:
```diff
@@ -0,0 +1,418 @@
+  };
+  struct Amplitude {
+    using type = double;
+    static constexpr OptionString help{"The amplitude of the gauge wave."};
```
should gauge -> Teukolsky?

moxcodes committed:
Remove unnecessary semicolon from BouncingBH

moxcodes commented:
Sorry @fmahebert ; I think those changes you suggested previously must have gotten lost in my git snafu. I've added fixups that add the changes again (hopefully I'll be able to squash properly this time).

fmahebert commented:
Heh no worries! LGTM, please squash (good luck 😛). Also it looks like you'll need to rebase away a conflict...

moxcodes commented:
Squashed (I believe correctly).
Cheers!
